### DISCUSSION TITLE
Getting an InvalidCastException for a simple TemplatedControl when the ControlTemplate includes a Label with a Target

### DISCUSSION THREAD

**Discussion by djeikyb** [AvaloniaUI/Avalonia#15915] [Q&A] [answered]
This is my first ever TemplatedControl and I was trying to do something a bit more complex, got stuck, and distilled my problem to this repo: https://github.com/djeikyb/LearnTemplateControls

My end goal is a control that can either be a labeled autocomplete input, or, tap a checkbox labeled "override" and it becomes a regular (labeled) text input. The idea is that by default you can choose from a (rather long) list of options, or if you know better, just type in what you know the input value ought to be.

I reduced this concept to a stack panel that wraps a label and a simple text box. But I'm getting an InvalidCastException when I add the target property to the label! In my ControlTemplate, how should I be working with the label's target prop so that the normal ergonomics of "tap the label to get the cursor into the label's text box" working?

Here's my templated control:

```cs
public class LabeledTextBox : TemplatedControl
{
    public static readonly StyledProperty<string?> LabelTextProperty =
        AvaloniaProperty.Register<LabeledTextBox, string?>(
            "LabelText");

    public string? LabelText
    {
        get => GetValue(LabelTextProperty);
        set => SetValue(LabelTextProperty, value);
    }
}
```

This template doesn't crash the app, but also doesn't associate the label with the text box:

```xaml
    <ControlTemplate x:Key="LabeledTextBoxNoTargetTemplate"
                     TargetType="local:LabeledTextBox">
        <StackPanel>
            <Label Content="{TemplateBinding LabelText}" />
            <TextBox Name="ManualEntry" />
        </StackPanel>
    </ControlTemplate>
```

But this one I thought would be correct, it tips the app over with the InvalidCastException right at launch:

```xaml
    <ControlTemplate x:Key="LabeledTextBoxWithTargetTemplate"
                     TargetType="local:LabeledTextBox">
        <StackPanel>
            <Label Content="{TemplateBinding LabelText}"
                   Target="ManualEntry" />
            <TextBox Name="ManualEntry" />
        </StackPanel>
    </ControlTemplate>
```

I suspect the ResolveByNameAttribute is at play here, I'm just not sure exactly how or what to do about it. It does seem potentially a bit dodgy to be using a static name prop in a template. Cause that could end up adding a bunch of elements to the tree, all with the same name, right? One for each time the control is used with that template?

**Comment by djeikyb** [comment]
So I added a third ControlTemplate. This one, to my surprise, doesn't fault at launch, even when I throw multiple local:LabeledTextBox into the window using this template:

```xaml
    <ControlTemplate x:Key="LabeledBelowTextBoxWithTargetTemplate"
                     TargetType="local:LabeledTextBox">
        <StackPanel>
            <TextBox Name="Asdf" />
            <Label Content="{TemplateBinding LabelText}"
                   Target="Asdf" />
        </StackPanel>
    </ControlTemplate>
```

Unfortunately, the label is on the bottom instead of on top.. 

<img width="269" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/794743/c81ea911-c778-461a-ace2-2c008664b0cb">

**Comment by stevemonaco** [comment, marked as answer]
Not entirely sure why this is, but `Target` bindings seem to work fine for me. You should also be using `ControlTheme` with style selectors instead of multiple top-level `ControlTemplate`s. You can use `IsVisible` on `Label` to toggle its visibility.

You can also set the `ControlTemplate` via selector, especially if you use conditional activators (like pseudoclasses). Usually you try to avoid restructuring the XAML tree so that's not necessary.

```xml
<ResourceDictionary
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="clr-namespace:LearnTemplateControls">
    <ControlTheme x:Key="{x:Type local:LabeledTextBox}" TargetType="local:LabeledTextBox">
        <Setter Property="Template">
            <ControlTemplate>
                <StackPanel>
                    <Label x:Name="PART_Label" Content="{TemplateBinding LabelText}" Target="{Binding #ManualEntry}" />
                    <TextBox x:Name="ManualEntry" />
                </StackPanel>
            </ControlTemplate>
        </Setter>
    </ControlTheme>
</ResourceDictionary>
```
Usage:
```xml
<local:LabeledTextBox LabelText="Some label text" />
```

You may be able to use `AutoCompleteBox` with `InnerLeftContent` containing a checkbox or toggleswitch. Might be a way to disable autocomplete. If so, then it's more straightforward.

**Reply by djeikyb** [reply]
Oh I see, you immediately went for

```xaml
<Label Target="{Binding #ManualEntry}" />
```

instead of:

```xaml
<Label Target="ManualEntry" />
```

that makes all the difference, but why doesn't `Target="ManualEntry"` work like I expect?

**Reply by djeikyb** [reply]
This is gonna come down to how the xaml is compiled and in what order are bindings created and resolved, isn't it lol

**Reply by djeikyb** [reply]
Ahhh I was hating the idea of having to ask for the template in my views every time I want to use the control. It was convenient for writing up my question tho

**Comment by djeikyb** [comment]
I stumbled onto a solution, but I'm not yet confident with my theory of how it works.

```xaml
    <ControlTemplate x:Key="LabelTargetUsesFunkyBindingSyntax"
                     TargetType="local:LabeledTextBox">
        <StackPanel>
            <Label Content="{TemplateBinding LabelText}"
                   Target="{Binding ElementName=foo}">
            </Label>
            <TextBox Name="foo" />
        </StackPanel>
    </ControlTemplate>
```

or — I think — equivalently:

```xaml
<Label Content="{TemplateBinding LabelText}"
    Target="{Binding #foo}">
```

I think the label's target prop can either (a) be passed a plain string, which should be the Name prop of another control, and then the ResolveByNameAttribute will work out a binding to that control (b) pass your own binding to the desired control.

It makes sense to me that passing my own binding clicks it together. I don't understand why passing the name as a literal string doesn't work as I originally described.

**Reply by stevemonaco** [reply]
Me neither. It goes through the XAML transformers part of the compilation process: https://github.com/AvaloniaUI/Avalonia/blob/9c89dee4dcf73c749dc823f383dbbeb497263e3a/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlResolveByNameMarkupExtensionReplacer.cs#L27

I don't know the area well, so it would take me quite awhile to debug.

**Comment by jp2masa** [comment]
> It does seem potentially a bit dodgy to be using a static name prop in a template. Cause that could end up adding a bunch of elements to the tree, all with the same name, right? One for each time the control is used with that template?

There is the concept of XAML namescopes, so this shouldn't be a problem (I don't think there are Avalonia docs for this, so see WPF docs [here](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/wpf-xaml-namescopes?view=netframeworkdesktop-4.8#xaml-namescopes-and-name-related-apis)).

Anyway, this seems to be a bug, likely where @stevemonaco points out (https://github.com/AvaloniaUI/Avalonia/discussions/15915#discussioncomment-9671402), so I think that an issue should be created.

**Reply by djeikyb** [reply]
> xaml namescopes

oh, nifty! thanks for the docs link

> seems to be a bug

~i can write it up in a bit~ https://github.com/AvaloniaUI/Avalonia/issues/15927
